bin/submitty_autograding_shipper.py

```diff
@@ -0,0 +1,417 @@
+#!/usr/bin/env python3
+
+import os
+import sys
+import time
+import signal
+import json
+import grade_items_logging
+import grade_item
+import shutil
+import contextlib
+import datetime
+from submitty_utils import glob
+import multiprocessing
+from submitty_utils import dateutils, glob
+import operator
+import paramiko
+import tempfile 
+# ==================================================================================
+# these variables will be replaced by INSTALL_SUBMITTY.sh
+NUM_GRADING_SCHEDULER_WORKERS_string = "__INSTALL__FILLIN__NUM_GRADING_SCHEDULER_WORKERS__"
+NUM_GRADING_SCHEDULER_WORKERS_int    = int(NUM_GRADING_SCHEDULER_WORKERS_string)
+
+AUTOGRADING_LOG_PATH="__INSTALL__FILLIN__AUTOGRADING_LOG_PATH__"
+SUBMITTY_DATA_DIR = "__INSTALL__FILLIN__SUBMITTY_DATA_DIR__"
+HWCRON_UID = "__INSTALL__FILLIN__HWCRON_UID__"
+INTERACTIVE_QUEUE = os.path.join(SUBMITTY_DATA_DIR, "to_be_graded_queue")
+
+# ==================================================================================
+def initialize(untrusted_queue):
+    """
+    Initializer function for all our processes. We get one untrusted user off our queue which
+    we then set in our Process. We cannot recycle the shipper process as else the untrusted user
+    we set for this process will be lost.
+
+    :param untrusted_queue: multiprocessing.queues.Queue that contains all untrusted users left to
+                            assign
+    """
+    multiprocessing.current_process().untrusted = untrusted_queue.get()
+
+# ==================================================================================
+# ==================================================================================
+def prepare_job(which_machine,which_untrusted,next_directory,next_to_grade):
+    # verify the hwcron user is running this script
+    if not int(os.getuid()) == int(HWCRON_UID):
+        grade_items_logging.log_message(message="ERROR: must be run by hwcron")
+        raise SystemExit("ERROR: the grade_item.py script must be run by the hwcron user")
+
+    # prepare the zip files
+    try:
+        autograding_zip_tmp,submission_zip_tmp = grade_item.prepare_autograding_and_submission_zip(which_machine,which_untrusted,next_directory,next_to_grade)
+        autograding_zip = os.path.join(SUBMITTY_DATA_DIR,"autograding_TODO",which_untrusted+"_autograding.zip")
+        submission_zip = os.path.join(SUBMITTY_DATA_DIR,"autograding_TODO",which_untrusted+"_submission.zip")
+        todo_queue_file = os.path.join(SUBMITTY_DATA_DIR,"autograding_TODO",which_untrusted+"_queue.json")
+
+        with open(next_to_grade, 'r') as infile:
+            queue_obj = json.load(infile)
+            queue_obj["which_untrusted"] = which_untrusted
+            queue_obj["which_machine"] = which_machine
+            queue_obj["ship_time"] = dateutils.write_submitty_date(microseconds=True)
+    except Exception as e:
```

Comment:
Be more specific about what Exception we're worried about here. Are we worried about `grade_item.prepare_autograding_and_submission_zip(which_machine,which_untrusted,next_directory,next_to_grade)` throwing an exception or just that opening `open(next_to_grade, 'r')` will fail?